polyval/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+//! POLYVAL is a GHASH-like universal hash over GF(2^128) useful for
+//! implementing AES-GCM-SIV (RFC 8452) as well as GMAC/AES-GCM.
+//!
+//! It can be thought of as the little endian equivalent of GHASH, which
+//! affords it a small performance advantage over GHASH when used on little
+//! endian architectures.
+//!
+//! By converting endianness on the way in and out, this means that POLYVAL can
+//! trivially be used to implement GHASH. This means POLYVAL can be used to
+//! implement GMAC, the Message Authentication Code used by AES-GCM.
+
+#![no_std]
+#![doc(html_logo_url = "https://raw.githubusercontent.com/RustCrypto/meta/master/logo_small.png")]
+
+extern crate byteorder;
+extern crate crypto_mac;
+#[cfg(feature = "zeroize")]
+extern crate zeroize;
+
+use byteorder::{BigEndian, ByteOrder, LittleEndian};
+use core::cmp::min;
+use core::marker::PhantomData;
+use crypto_mac::generic_array::{typenum::U16, GenericArray};
+use crypto_mac::{Mac, MacResult};
+#[cfg(feature = "zeroize")]
+use zeroize::Zeroize;
+
+/// The POLYVAL universal hash function
+pub type Polyval = PolyvalCore<LittleEndian>;
+
+/// The GHASH universal hash function
+pub type GHash = PolyvalCore<BigEndian>;
+
+/// POLYVAL: GHASH-like universal hash over GF(2^128).
+///
+/// From RFC 8452, Appendix A:
+///
+/// > The Relationship between POLYVAL and GHASH
+/// >
+/// > GHASH and POLYVAL both operate in GF(2^128), although with different
+/// > irreducible polynomials: POLYVAL works modulo x^128 + x^127 + x^126 +
+/// > x^121 + 1 and GHASH works modulo x^128 + x^7 + x^2 + x + 1.  Note
+/// > that these irreducible polynomials are the "reverse" of each other.
+///
+/// Using this property, we are able to implement both the POLYVAL and GHASH
+/// functions in terms of POLYVAL simply by being generic over the byte order.
+/// However, POLYVAL has the advantage of being optimized for natively
+/// little-endian CPUs.
+#[repr(align(16))]
+#[derive(Clone)]
+pub struct PolyvalCore<B: ByteOrder> {
+    h: u128,
+    s: u128,
+    leftover: usize,
+    buffer: [u8; 16],
+    o: PhantomData<B>,
+}
+
+const XMM_MASK: u128 = 0x0001_c200_0000_0000_0000;
+
+// TODO(tarcieri): POLYVAL is a universal hash function, not a MAC.
+// It can be used to implement a MAC (e.g. GMAC), but should probably use
+// a different trait than `crypto_mac::Mac`.
```

Comment:
Should really *not* use `Mac` here, IMO.

Reply:
Removed the `Mac` impl. It now exposes a `Mac`-like API, but one that only operates in term of `FieldElement`-sided blocks.